tv_apps/smart-home/js/home.js

```diff
@@ -187,10 +195,148 @@
+                  () => this.exitEditMode());
+              }
+            } else {
+              // Don't agree to delete. Nothing happens so let's just exit.
+              this.exitEditMode();
+            }
+          });
+      });
+
+      this.editFolderMenuItem.addEventListener('click', () => {
+        this._cardPicker.show(this.cardScrollable.currentItem);
+      });
+    },
+
+    enableContextMenu: function () {
+      var focus = this.spatialNavigator.getFocusedElement();
+      if (this.mode === '' &&
+          (focus === this.cardScrollable || focus === this.folderScrollable)) {
+        this.mainSection.setAttribute('contextmenu', this.contextmenuElem.id);
+      }
+    },
+
+    disableContextMenu: function () {
+      this.mainSection.setAttribute('contextmenu', '');
+    },
+
+    updateContextMenu: function (itemElem) {
+      var type = itemElem.getAttribute('app-type');
+      if (this.contextmenuElem.dataset.currentAppType == type) {
+        return;
+      }
+      this.contextmenuElem.dataset.currentAppType = type;
+
+      // Clean up first
+      for (var i = this.contextmenuElem.childNodes.length - 1; i >= 0; --i) {
+        this.contextmenuElem.removeChild(this.contextmenuElem.childNodes[i]);
+      }
+
+      this.contextmenuElem.appendChild(this.moveMenuItem);
+
+      switch (type) {
+        case 'folder':
+          this.contextmenuElem.appendChild(this.editFolderMenuItem);
+          break;
+
+        case 'tv':
+        case 'app':
+        case 'appbookmark':
+          this.contextmenuElem.appendChild(this.renameMenuItem);
+          this.contextmenuElem.appendChild(this.removeMenuItem);
+          break;
+      }
+    },
+
+    enterEditMode: function (onEnter) {
+      return new Promise((resolve, reject) => {
+        var scrollable = this.spatialNavigator.getFocusedElement();
+        this._currentEditedCardId = scrollable.currentItem.dataset.cardId;
+
+        if (scrollable === this.folderScrollable) {
+          // Folder would be cleaned then rebuilt when entering edit mode.
+          // So if we are editing card under folder,
+          // we have to find and make card being edited refocused
+          // after folder is rebuilt.
+          this.once('built-folder-list', () => {
+            var item = this.folderListElem.querySelector(
+              '.app-button[data-card-id="' + this._currentEditedCardId + '"]');
+            this.spatialNavigator.focus(scrollable);
+            scrollable.focus(item);
+            resolve();
+          });
+        } else {
+          resolve();
+        }
+
+        this.cleanFolderScrollable();
+        this.edit.toggleEditMode();
+        // XXX: Reset card filter when entering edit mode
+        this.filterManager.resetFilter();
+      });
+    },
 
-        that.editFolderElem.addEventListener(
-                                        'click', that.onEditFolder.bind(that));
+    exitEditMode: function () {
+      this.edit.once('exit-edit-mode', () => {
```

Comment:
I guess you can completely remove these end-of-switching-mode-event and just do them synchronously.